src/main/java/com/eerussianguy/blazemap/feature/maps/AtlasExporter.java

```diff
@@ -0,0 +1,144 @@
+package com.eerussianguy.blazemap.feature.maps;
+
+import com.eerussianguy.blazemap.api.BlazeRegistry;
+import com.eerussianguy.blazemap.api.maps.Layer;
+import com.eerussianguy.blazemap.api.maps.MapType;
+import com.eerussianguy.blazemap.api.maps.TileResolution;
+import com.eerussianguy.blazemap.api.util.RegionPos;
+import com.eerussianguy.blazemap.engine.BlazeMapAsync;
+import com.eerussianguy.blazemap.engine.StorageAccess;
+import com.eerussianguy.blazemap.engine.client.BlazeMapClientEngine;
+import com.eerussianguy.blazemap.engine.client.LayerRegionTile;
+import com.eerussianguy.blazemap.util.Colors;
+import com.mojang.blaze3d.platform.NativeImage;
+import net.minecraft.client.Minecraft;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.Level;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AtlasExporter {
+    private static Task task = null;
+
+    public static synchronized void export(Task task) {
+        if(AtlasExporter.task == null) {
+            AtlasExporter.task = task;
+            BlazeMapAsync.instance().clientChain.runOnDataThread(AtlasExporter::export);
+        }
+    }
+
+    public static synchronized Task getTask() {
+        return task;
+    }
+
+    private static void export() {
+        TileResolution resolution = TileResolution.FULL;
+        StorageAccess.Internal storage = BlazeMapClientEngine.getDimensionStorage(task.dimension);
+        AtlasInfo atlasInfo = new AtlasInfo();
+        NativeImage atlas = constructAtlas(storage, resolution, atlasInfo);
+
+        try(atlas) {
+            var layerKeys = task.map.value().getLayers();
+            for(var layerKey : layerKeys) {
+                if(!task.layers.contains(layerKey)) continue;
+                File folder = storage.getMipmap(layerKey.location, ".", resolution);
+                for(int regionX = atlasInfo.atlasStartX; regionX <= atlasInfo.atlasEndX; regionX++) {
+                    for(int regionZ = atlasInfo.atlasStartZ; regionZ <= atlasInfo.atlasEndZ; regionZ++) {
+                        File file = new File(folder, LayerRegionTile.getImageName(new RegionPos(regionX, regionZ)));
+                        if(!file.exists()) continue;
+                        NativeImage tile = NativeImage.read(Files.newInputStream(file.toPath()));
```

Comment:
You probably want to hold the region's file read lock while performing this action to prevent corruption from another thread writing to this region at the same time

Reply:
For a future refactor, this should probably directly use `LayerRegionTile::tryLoad`, though not for now as that updates some static vars that need to be separated out first.

Reply:
We should lock the file itself.
Trying to lock in code will require going though the engine's cache, and that's exactly what I wanted to avoid because of all the other performance implications.

Reply:
Yes, we want to lock the file itself, but we do have to lock the file _using_ code. This has nothing to do with the engine's cache. But anyways, can discuss this for later. This is fine for now.

Reply:
I fixed everything else, we'll come back to this later